cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/InvalidPointerToDereference.qll

```diff
@@ -66,11 +66,14 @@
  * module. Since the node we are tracking is not necessarily _equal_ to the pointer-arithmetic instruction, but rather satisfies
  * `node.asInstruction() <= pai + deltaDerefSourceAndPai`, we need to account for the delta when checking if a guard is sufficiently
  * strong to infer that a future dereference is safe. To do this, we check that the guard guarantees that a node `n` satisfies
- * `n < node + k` where `node` is a node we know is equal to the value of the dereference source (i.e., it satisfies
- * `node.asInstruction() <= pai + deltaDerefSourceAndPai`) and `k <= deltaDerefSourceAndPai`. Combining this we have
- * `n < node + k <= node + deltaDerefSourceAndPai <= pai + 2*deltaDerefSourceAndPai` (TODO: Oops. This math doesn't quite work out.
- * I think this is because we need to redefine the `BarrierConfig` to start flow at the pointer-arithmetic instruction instead of
- * at the dereference source. When combined with TODO above it's easy to show that this guard ensures that the dereference is safe).
+ * `n < node + k` where `node` is a node such that `node <= pai`. Thus, we know that any node `m` such that `m <= n + delta` where
+ * `delta + k <= 0` will be safe because:
+ * ```
+ * m <= n + delta
+ *   <  node + k + delta
+ *   <= pai + k + delta
+ *   <= pai
```

Comment:
Hmm... I don't think `k` and `delta` are necessarily both positive? In fact, `k` is probably negative in most sensible code since you'll have a check such as
```cpp
int* end = ... // construct invalid pointer
if(my_ptr < end) { // k is -1 here since range analysis infers that `my_ptr <= end - 1`
  *my_ptr = 0;
}
```
And so the requirement that `delta + k <= 0` can be read as "`k` must be sufficiently negative (i.e., the `my_ptr < end` guard must be strong enough) to guarantee that even if the pai constructed a pointer that's "off-by-1000" elements the k must be so negative that the guard ensures that `my_ptr` is safe to dereference".

Does that make sense?

Reply:
So this checks out.